libsql_client/dbapi2/types.py

```
@@ -0,0 +1,1423 @@
+        "INSERT", "UPDATE", "DELETE", "REPLACE",
+    }
+
+    @overload
+    def __init__(self, sql: str) -> None:
+        ...
+
+    @overload
+    def __init__(self, sql: str, _allow_instantiation: "Connection") -> None:
+        ...
+
+    def __init__(self, *args: object, **kwargs: object) -> None:
+        # NOTE do this to enable test_dbapi.py
+        # ModuleTests.test_disallow_instantiation(), it will call
+        # the connection to get a Statement, but the statement itself
+        # must not be instantiated directly.
+        # Test instantiates the result of: type(cx("select 1"))
+        con = kwargs.get("_allow_instantiation")
+        if not isinstance(con, Connection):
+            # matches check_disallow_instantiation() body
+            tp = self.__class__
+            mod = tp.__module__
+            name = tp.__name__
+            qualname = f"{mod}.{name}"
+            msg = f"cannot create '{qualname}' instances"
+            raise TypeError(msg)
+
+        assert len(args) == 1
+        sql = args[0]
+        if not isinstance(sql, str):
+            raise TypeError("sql must be a str")
+
+        if "\0" in sql:
+            raise ProgrammingError("the query contains a null character")
+        if sys.version_info[:2] >= (3, 11):
+            if len(sql) > con.getlimit(SQLITE_LIMIT_SQL_LENGTH):
+                raise DataError("query string is too large")
+
+        self.sql = sql
+        # see pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L32
+        # however we need to parse the statement with iter_sql_statements()
+        # as we can't use sqlite3_prepare_v2()
+        itr = iter(iter_sql_statements(sql))
+        self.tokens = next(itr)
+        assert self.tokens
+        next_stmt = next(itr, None)
+        if next_stmt and next_stmt[0].upper() != "END":
+            msg = "You can only execute one statement at a time."
+            raise ProgrammingError(msg)
```

Why we ignore this error when the second statement is `END`?

---

they have a test with `CREATE TRIGGER ... BEGIN ... ; END`. From their C code, this is handled as a single thing by https://www.sqlite.org/c3ref/prepare.html

that's what they call "a statement". They don't do the parsing themselves... I can't call C to check, so that's the best I can do to keep the same behavior.

---

The `CREATE TRIGGER` statement can contain multiple statements before the `END`, i.e. `CREATE TRIGGER ... BEGIN SELECT 1; SELECT 2; SELECT 3; END`.
https://www.sqlite.org/lang_createtrigger.html

I think that we should either not support `CREATE TRIGGER` at all, or try to parse it correctly by checking for statements that begin with tokens `CREATE`, `TRIGGER` and looking for the matching `END`. Supporting only CREATE TRIGGERs that contain exactly one statement will be confusing.

---

yeah, I'm not even sure Turso supports those, I recall having to disable the test case for that and it wasn't related to my part. Do you know?

Alternatively I though of a hack to simplify this part, but it's kinda ugly: I can keep a singleton `connection(":memory:")` and call it (`__call__()`), as you see in my code, which matches the C, all it does is to call prepare on the statement and evaluate it...